packages/components/src/Button/Button.tsx

```diff
@@ -1,83 +1,115 @@
-    ariaHaspopup,
-    ariaExpanded,
-    ariaLabel,
-    disabled = false,
-    external,
-    id,
-    name,
-    onClick,
-    onMouseDown,
-    role,
-    value,
-    submit,
-    to,
-    url,
-    UNSAFE_className = {},
-    UNSAFE_style = {},
-    children,
-  } = props;
-
-  const { combined } = useButtonStyles(props);
-
-  const buttonType: HTMLButtonType = submit ? "submit" : "button";
-
-  const buttonClassNames = classnames(combined, UNSAFE_className.container);
-
-  const tagProps = {
-    className: buttonClassNames,
-    disabled,
-    id,
-    style: UNSAFE_style.container,
-    ...(submit && { name, value }),
-    ...(!disabled && { href: url }),
-    ...(!disabled && { onClick: onClick }),
-    ...(!disabled && { onMouseDown: onMouseDown }),
-    ...(external && { target: "_blank" }),
-    ...(url === undefined && to === undefined && { type: buttonType }),
-    "aria-controls": ariaControls,
-    "aria-haspopup": ariaHaspopup,
-    "aria-expanded": ariaExpanded,
-    "aria-label": ariaLabel,
-    role: role,
-  };
-
-  const buttonInternals = children || <ButtonContent {...props} />;
-
-  if (to) {
+// Define the possible ref types
+type ButtonRefElement = HTMLButtonElement | HTMLAnchorElement;
+
+const ButtonComponent = forwardRef<ButtonRefElement, ButtonProps>(
+  (props, ref) => {
+    const { size } = props;
+
     return (
-      <Link {...tagProps} to={to}>
-        {buttonInternals}
-      </Link>
+      <ButtonProvider size={size}>
+        <ButtonWrapper {...props} ref={ref} />
+      </ButtonProvider>
     );
-  }
+  },
+);
+ButtonComponent.displayName = "Button";
+
+const ButtonWrapper = forwardRef<ButtonRefElement, ButtonProps>(
+  // eslint-disable-next-line max-statements
+  (props, ref) => {
+    const {
+      ariaControls,
+      ariaHaspopup,
+      ariaExpanded,
+      ariaLabel,
+      disabled = false,
+      external,
+      id,
+      name,
+      onClick,
+      onMouseDown,
+      role,
+      value,
+      submit,
+      to,
+      url,
+      UNSAFE_className = {},
+      UNSAFE_style = {},
+      children,
+    } = props;
+
+    const { combined } = useButtonStyles(props);
+
+    const buttonType: HTMLButtonType = submit ? "submit" : "button";
+
+    const buttonClassNames = classnames(combined, UNSAFE_className.container);
+
+    // Base props common to all element types
+    const commonProps = {
+      className: buttonClassNames,
+      disabled,
+      id,
+      style: UNSAFE_style.container,
+      ...(!disabled && { onClick: onClick }),
+      ...(!disabled && { onMouseDown: onMouseDown }),
+      "aria-controls": ariaControls,
+      "aria-haspopup": ariaHaspopup,
+      "aria-expanded": ariaExpanded,
+      "aria-label": ariaLabel,
+      role: role,
+    };
+
+    const buttonInternals = children || <ButtonContent {...props} />;
+
+    if (to) {
+      // Props specific to Link
+      const linkProps = {
+        ...commonProps,
+        to: to as LinkProps["to"],
+      };
```

Comment:
I've omitted the ref here since `to` is deprecated but maybe this is just making it more confusing than it's worth since the type won't tell you that in this scenario the ref won't be attached to anything. it'll appear valid but do nothing.